### PR TITLE
documentation: mention additionalArgs

### PR DESCRIPTION
**Pull request by jan--f**
## Description

For lists strategic merge patch can have surprising results where a list is replaced instead of added to. The operator offers workarounds for that, which should be explicitly mentioned.

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [x] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

### CODE REVIEW

**Comment by jan--f** [comment]
Not sure if an experimental flag is acceptable for a doc section. It was tricky to find something more benign that isn;t covered by a CRD field yet.

File: Documentation/user-guides/strategic-merge-patch.md [hunk 1]
```diff
@@ -99,3 +99,22 @@ spec:
     - sleep
     - "3600"
 ```
+
+## How to inject additional cli arguments into the prometheus pods
+
+The following manifest injects an additional cli argument in the default
+Prometheus argument list. Not the use of `additionalArgs` in this example.
+Using `args` directly would overwrite the args list completely, including the
```
Comment (machine424) [inline Documentation/user-guides/strategic-merge-patch.md:107]:
```suggestion
Using `.spec.containers[*].args` instead would overwrite the container's args list completely, including the
```

File: Documentation/user-guides/strategic-merge-patch.md [hunk 2]
```diff
@@ -99,3 +99,22 @@ spec:
     - sleep
     - "3600"
 ```
+
+## How to inject additional cli arguments into the prometheus pods
+
+The following manifest injects an additional cli argument in the default
+Prometheus argument list. Not the use of `additionalArgs` in this example.
```
Comment (machine424) [inline Documentation/user-guides/strategic-merge-patch.md:106]:
```suggestion
Prometheus argument list. Note the use of `.spec.additionalArgs` in this example.
```

File: Documentation/user-guides/strategic-merge-patch.md [hunk 3]
```diff
@@ -99,3 +99,22 @@ spec:
...
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: Prometheus
+metadata:
+  name: additional-arguments
+spec:
+  containers:
+  - name: prometheus
+    additionalArgs:
```
Comment (machine424) [inline Documentation/user-guides/strategic-merge-patch.md:118]: I thought it was directly under `spec`?

File: Documentation/user-guides/strategic-merge-patch.md [hunk 4]
```diff
@@ -99,3 +99,22 @@ spec:
...
+    - '--scrape.timestamp-tolerance=15ms'
```
Comment (machine424) [inline Documentation/user-guides/strategic-merge-patch.md:119]: maybe just `--arg=value`?

File: Documentation/user-guides/strategic-merge-patch.md [hunk 5]
```diff
@@ -99,3 +99,22 @@ spec:
     - sleep
     - "3600"
 ```
+
+## How to inject additional cli arguments into the prometheus pods
```
Comment (simonpasquier) [inline Documentation/user-guides/strategic-merge-patch.md:103]: (nit)
```suggestion
## How to inject additional CLI arguments into the prometheus container
```

File: Documentation/user-guides/strategic-merge-patch.md [hunk 6]
```diff
@@ -99,3 +99,22 @@ spec:
...
+
+The following manifest injects an additional cli argument in the default
```
Comment (simonpasquier) [inline Documentation/user-guides/strategic-merge-patch.md:105]:
```suggestion
The following manifest injects an additional CLI argument in the default
```

File: Documentation/user-guides/strategic-merge-patch.md [hunk 7]
```diff
@@ -99,3 +99,23 @@ spec:
...
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: Prometheus
+metadata:
+  name: additional-arguments
+spec:
+  containers:
+  - name: prometheus
+    additionalArgs:
+  - "--scrape.timestamp-tolerance=15ms"
```
Comment (simonpasquier) [inline Documentation/user-guides/strategic-merge-patch.md:120]:
```suggestion
  additionalArgs:
  - name: "scrape.timestamp-tolerance"
  - value: "15ms"
```

File: Documentation/user-guides/strategic-merge-patch.md [hunk 8]
```diff
@@ -99,3 +99,22 @@ spec:
     - sleep
     - "3600"
 ```
+
+## How to inject additional CLI arguments into the prometheus pods
```
Comment (ArthurSens) [inline Documentation/user-guides/strategic-merge-patch.md:103]:
```suggestion
## How to inject additional CLI arguments into the prometheus container
```

**Review by ArthurSens — COMMENTED** [review]
Besides that LGTM!

File: Documentation/user-guides/strategic-merge-patch.md [hunk 9]
```diff
@@ -99,3 +99,22 @@ spec:
     - sleep
     - "3600"
 ```
+
+## How to inject additional CLI arguments into the prometheus container
+
+The following manifest injects an additional CLI argument in the default
+Prometheus argument list. Note the use of `.spec.additionalArgs` in this
+example.
+Using `.spec.containers[*].args` directly would instead overwrite the containers
```
Comment (simonpasquier) [inline Documentation/user-guides/strategic-merge-patch.md:108]:
```suggestion
Using `.spec.containers[*].args` directly would instead overwrite the container's
```

**Review by simonpasquier — APPROVED** [review]

**Comment by ArthurSens** [comment]
Automerge won't work because it's only touching .md file 😬

**Comment by nicolastakashi** [comment]
> Sorry for deviating from the subject of this PR; it reminded me of a subject we [discussed recently during our office hours](https://docs.google.com/document/d/1-fjJmzrwRpKmSPHtXN5u6VZnn39M28KqyQGBEJsqUOk/edit#heading=h.w6vs0qd5jbgf).
> 
> We wanted to improve documentation for the strategic merge patch, while also working on updating the behavior here to not completely override cmdArgs or environment variables.
> 
> @simonpasquier @nicolastakashi, seeing that we have a field for additional cmd args, what do you think about just adding a field of additional environment variables and leaving the merge patch as it is? I'm also concerned that changing the behavior of the merge patch would require Prometheus API 2.0

Sounds reasonable to me TBH

**Comment by simonpasquier** [comment]
> seeing that we have a field for additional cmd args, what do you think about just adding a field of additional environment variables and leaving the merge patch as it is? I'm also concerned that changing the behavior of the merge patch would require Prometheus API 2.0

I wouldn't necessarily consider this to be a breaking change if the CRD struct doesn't change but yes it might be disturbing and trigger the opposite request (e.g. "I want to overwrite all env variables and it's not possible anymore").

In principle, I'm ok for a `.spec.additionalEnv` field but the API is going to be more complicated than for args because you can get the envvar value from different ways (e.g. static, field ref, secret ref, ...).

**Comment by simonpasquier** [comment]
In fact I tested and environment variables are "properly" merged in the resulting statefulset. This is because the `Env` field in the `v1.Container` struct has the right `patchMergeKey` and `patchStrategy` annotations. I believe that it will make a good example for the docs.

**Comment by ArthurSens** [comment]
> In fact I tested and environment variables are "properly" merged in the resulting statefulset. This is because the `Env` field in the `v1.Container` struct has the right `patchMergeKey` and `patchStrategy` annotations. I believe that it will make a good example for the docs.

Oh, that's news to me 😮 

Does it means that the $SHARD env is not overwritten when doing a strategic merge patch?

**Comment by simonpasquier** [comment]
Indeed

**Pull request merged**